### PR TITLE
Make %dirname behave like dirname (3)

### PR DESCRIPTION
**Pull request by ffesti**
Strip trailing "/"
Keep "." and ".." untouched
Keep root dir for files in the root dir
Return "." for files without directory and empty strings

See man 3 dirname

Resolves: 2928

### CODE REVIEW

File: rpmio/macro.c [hunk 1]
```diff
@@ -1306,8 +1306,22 @@ static void doFoo(rpmMacroBuf mb, rpmMacroEntry me, ARGV_t argv, size_t *parsed)
...
+		buf[l-1] = '\0';
+	    if ((b = strrchr(buf, '/')) != NULL)
+		if (b == buf)
+		    /* keep root dir */
+		    buf[1] = '\0';
+		else
+		    *b = '\0';
+	    else
+		strcpy(buf, ".");
+	}
```
Comment (pmatilai) [inline rpmio/macro.c:1324]: Um, why not call dirname() instead?

Comment (ffesti) [reply rpmio/macro.c]: Good question. Why didn't the code used that in the first place...?

**Comment by pmatilai** [comment]
Right, I don't know why these aren't calling the respective stdlib functions to begin with. It's not like it's expensive or anything.

**Comment by mlschroe** [comment]
See https://pubs.opengroup.org/onlinepubs/9699919799/functions/basename.html for testcases ;-)

File: rpmio/macro.c (hunk 1 above)
Comment (pmatilai) [reply rpmio/macro.c]: It's a good question. You have to go pretty far back (commit 94f5fbeec363059ca9eb986ee294ffcdf20bcb04) and come back none the wiser :smile:

File: rpmio/macro.c [hunk 2]
```diff
@@ -1306,9 +1307,9 @@ static void doFoo(rpmMacroBuf mb, rpmMacroEntry me, ARGV_t argv, size_t *parsed)
 	    b++;
     } else if (rstreq("dirname", me->name)) {
 	buf = xstrdup(argv[1]);
-	if ((b = strrchr(buf, '/')) != NULL)
-	    *b = '\0';
-	b = buf;
+	b = xstrdup(dirname(buf));
```
[the remark this answers is not in the document]
Comment (ffesti) [reply rpmio/macro.c]: Yeah, I figured that myself but it took me a while to figure out that the simplest solution actually works. dirname and basename are a bit weird memory handling wise.

[the remark this answers is not in the document]
Comment (pmatilai) [reply rpmio/macro.c]: They are indeed, and then there's glibc version of basename() which differs in that it takes a const char * argument and relying on that is incompatible and ....

**Comment by pmatilai** [comment]
Maybe the basename() case should also have some more advanced tests than the one that is there to match the added dirname cases, see @mlschroe's link. Other than that, looks fine now.

**Review by pmatilai — APPROVED** [review]

**Pull request merged**